testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java

```diff
@@ -364,6 +370,106 @@ else if (getHiveVersionMajor() == 3 && isFormat.test("orc")) {
                 .buildOrThrow();
     }
 
+    protected void doTestHiveCoercionWithDifferentTimestampPrecision(HiveTableDefinition tableDefinition)
+    {
+        String tableName = mutableTableInstanceOf(tableDefinition).getNameInDatabase();
+
+        // Insert all the data with nanoseconds precision
+        setHiveTimestampPrecision(NANOSECONDS);
+        onTrino().executeQuery(
+                """
+                    INSERT INTO %s VALUES
+                    (TIMESTAMP '2121-07-15 15:30:12.123499', 1),
+                    (TIMESTAMP '2121-07-15 15:30:12.123500', 1),
+                    (TIMESTAMP '2121-07-15 15:30:12.123501', 1),
+                    (TIMESTAMP '2121-07-15 15:30:12.123499999', 1),
+                    (TIMESTAMP '2121-07-15 15:30:12.123500000', 1),
+                    (TIMESTAMP '2121-07-15 15:30:12.123500001', 1)
+                    """.formatted(tableName));
+
+        onHive().executeQuery(format("ALTER TABLE %s CHANGE COLUMN timestamp_to_varchar timestamp_to_varchar STRING", tableName));
+
+        for (HiveTimestampPrecision hiveTimestampPrecision : HiveTimestampPrecision.values()) {
+            setHiveTimestampPrecision(hiveTimestampPrecision);
+            assertThat(onTrino().executeQuery("SHOW COLUMNS FROM " + tableName).project(1, 2)).containsExactlyInOrder(
+                    row("timestamp_to_varchar", "varchar"),
+                    row("id", "bigint"));
+
+            List<String> allColumns = ImmutableList.of(
+                    "timestamp_to_varchar",
+                    "id");
+
+            // For Trino, remove unsupported columns
+            List<String> trinoReadColumns = removeUnsupportedColumnsForTrino(allColumns, tableName);
+            Map<String, List<Object>> expectedTinoResults = Maps.filterKeys(
+                    expectedRowsForEngineProvider(Engine.TRINO, hiveTimestampPrecision),
+                    trinoReadColumns::contains);
+
+            String trinoReadQuery = format("SELECT %s FROM %s", String.join(", ", trinoReadColumns), tableName);
+            assertQueryResults(Engine.TRINO, trinoReadQuery, expectedTinoResults, trinoReadColumns, 6, tableName);
+
+            List<String> hiveReadColumns = removeUnsupportedColumnsForHive(allColumns, tableName);
+            Map<String, List<Object>> expectedHiveResults = Maps.filterKeys(
+                    expectedRowsForEngineProvider(Engine.HIVE, hiveTimestampPrecision),
+                    hiveReadColumns::contains);
+
+            String hiveSelectQuery = format("SELECT %s FROM %s", String.join(", ", hiveReadColumns), tableName);
+            assertQueryResults(Engine.HIVE, hiveSelectQuery, expectedHiveResults, hiveReadColumns, 6, tableName);
+        }
+    }
+
+    protected Map<String, List<Object>> expectedRowsForEngineProvider(Engine engine, HiveTimestampPrecision hiveTimestampPrecision)
```

Comment:
we should return same results as hive, not "somewhat similar"

the column is varchar. if hive returns `2121-07-15 15:30:12.123499999` (29 characters), Trino should also return `2121-07-15 15:30:12.123499999` (29 characters).

remove this method

Reply:
Hive timestamp doesn't have the concept of precision - how do we handle them in trino

Reply:
> So should it be irrespective of the timestamp precision specified as a part session or config property ?

i believe so

> Hive timestamp doesn't have the concept of precision - 

i know. they are always nanoseconds precision

> how do we handle them in trino

regular timestamps get truncated to selected precision for performance reasons
for columns that were timestamps and are now varchars, we don't have these performance considerations and we should be compatible with hive.